scripts/determine_test_files_to_run.py

```diff
@@ -0,0 +1,229 @@
+The specificity of the algorithm can be tweaked through the `depth` argument (how many layers out do you want to traverse).
+
+While this script does not provide as much coverage as a traditional test run, the fact that it traverses GE's internal dependency
+graph layer by layer to determine the most relevant files allows us to maintain high coverage (all while improving performance).
+
+If a key file like `data_context.py` is changed, the vast majority of the test suite will run because the corresponding node in
+the dependency graph has so many ingoing and outgoing connections. This treatment of "high traffic" areas allow us to keep a watchful eye
+over the most important and used parts of our codebase.
+
+"""
+import ast
+import glob
+import os
+import subprocess
+from collections import namedtuple
+from typing import Dict, List, Tuple
+
+Import = namedtuple("Import", ["source", "module", "name", "alias"])
+
+
+def get_changed_files() -> Tuple[List[str], List[str]]:
+    """
+    Standard git diff against `develop` to determine list of changed files.
+    Filters out anything that isn't in `great_expectations/`.
+    """
+    process = subprocess.run(
+        ["git", "diff", "HEAD", "origin/develop", "--name-only"], stdout=subprocess.PIPE
+    )
+    files = [f.decode("utf-8") for f in process.stdout.splitlines()]
+    test_files = [f for f in files if f.startswith("tests")]
+    source_files = [f for f in files if f.startswith("great_expectations")]
+    return source_files, test_files
+
+
+def parse_imports(path: str) -> List[Import]:
+    """
+    Traverses a file using AST and determines relative imports (from within GE)
+
+    Parses both:
+      * from great_expectations.x.y.z import ...
+      * from great_expectation import ...
+    """
+    imports = []
+    with open(path) as f:
+        root = ast.parse(f.read(), path)
+
+    for node in ast.walk(root):
+        if isinstance(node, ast.Import):
+            module = []
+        elif isinstance(node, ast.ImportFrom) and node.module is not None:
+            module = node.module.split(".")
+        else:
+            continue
+
+        for n in node.names:
+            imp = Import(path, module, n.name.split("."), n.asname)
+            imports.append(imp)
+
+    return imports
+
+
+def get_import_paths(imports: List[Import]) -> List[str]:
+    """
+    Takes a list of Imports and determines the relative path to each source file or module.
+
+    Imports of type:
+      * `from great_expectations.x.y.z import ...` originate from `great_expectations/x/y/z.py`.
+      * `from great_expectations import ...` must be evaluated on a case by case basis.
+         Thankfully, there are only two edge cases we need to consider (DataContext and exceptions).
+
+    If it is the case that an Import is pointing to a module or directory, we play it safe
+    and add all files from that directory to ensure a high level of coverage.
+    """
```

Comment:
This uses AST to parse the code and make an educated guess about the source location of the import. Some testing shows that this is very accurate but the highest level of accuracy comes from actually running the underlying code.

A note for myself but look into `__import__("great_expectations.x.y.z").__file__` as a way to more accurately do this (at the expense of performance)